### ISSUE TITLE
Developers should be able to work on the welcome screen in our Rust codebase

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4152]
This task is automatically imported from the old Task Issue Board and it was originally created by Michael Mauderer.
Original issue is [here](https://www.pivotaltracker.com/story/show/180312485).
<hr />

## Why
This was meant to be implemented in Rust but was done in JS under time pressure. Having it in rust will lessen the maintenance burden on it in the future. 

**As** developer
**I want** to make code changes to the welcome screen in Rust
**So that** I benefit from the Rust tooling, and the welcome screen benefits from increased performance and code base integration. 

## Acceptance Criteria

```gherkin
Scenario: 
Given I work on the codebase
When I make changes to the welcome screen
Then I don't have to touch JS
```

## Resources
https://enso.youtrack.cloud/issue/IDE-3133

## Tasks:
- [X] Welcome Screen must be opened after download progress spinner
- [X] Welcome Screen must display list of available projects
- [X] Welcome Screen must allow to create new projects
- [X] Welcome Screen must allow to open projects and templates

## Blockers:
[#180388845](https://www.pivotaltracker.com/story/show/#180388845) resolved
[#180388817](https://www.pivotaltracker.com/story/show/#180388817) resolved
[#180388874](https://www.pivotaltracker.com/story/show/#180388874) resolved

## Comments:
**Ilya Bogdanov** reports a new **STANDUP** for today (11/16/2021):

**Today:** working on integration level and interactive features of welcome screen. Implemented creating and opening projects, automatic view switch. Opening templates is not implemented yet because it requires updating protocol definition (it misses template parameter). It should be finished by 2021-11-18.

**Next Day:** Next day I will be working on the [#180312485 task](https://www.pivotaltracker.com/story/show/180312485). continue working on welcome screen. The task should probably be finished tomorrow, but there are a number of things that need to be fixed/implemented: lock position on panning (add new layer), refactor added source code, test if passing project param skips welcome screen, write documentation, rebase on latest master, prepare PR (Enso Bot - Nov 16, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for today (11/17/2021):

**Today:** Adding new panel layer, refactoring and rebasing welcome screen code, splitting it to small PRs. Created two of PRs It should be finished by 2021-11-18.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180312485). continue working on welcome screen, hopefully create rest of PRs (Enso Bot - Nov 17, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for yesterday (2021-11-18):

**Today:** addressing review questions and comments, preparing documentation for welcome screen, moving CSS styles from .css file to Rust It should be finished by 2021-11-26.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180312485). day off for Friday, then continue working on welcome screen (Enso Bot - Nov 19, 2021)
<hr />
**Ilya Bogdanov** reports a new **🔴 DELAY** for today (2021-11-18):

**Summary:** There is **4 days** delay in implementation of the [#180312485 task](https://www.pivotaltracker.com/story/show/180312485).
It **will cause 4 days delay** for the delivery of this sprint.

not sure if I understand Sprint Delay field correctly, the task should be finished before the sprint ends even with such a delay.

**Delay Cause:** There were multiple causes for the delay. First of all, I got sick at the end of the week, which slowed progress on task. Secondly, we decided to split the task into several sub-tasks and deliver each subtask in separate PR (targeted at a temporary feature-branch). This led to splitting commits into multiple branches and refactoring the code for reviewers sake. Thirdly, I underestimated the cost of preparing PRs and resolving a discussions that arose during the review. I also decided to move CSS-styles definitions to Rust code from CSS files for better code structure, which took additional time.

**Possible solutions:** I need to report delays (brace for pun!) without a delay and better estimate tasks by splitting them into subtasks early. Perhaps additional "research" chore should be made before going straight into the implementation. Splitting to subtasks early would also help with design discussions, which in turn should reduce the time needed for review. (Enso Bot - Nov 21, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for yesterday (2021-11-23):

**Today:** working on resolving review comments for latest Welcome Screen PR, most of them are resolved by now, some discussions are to be done. It should be finished by 2021-11-26.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180312485). continue pushing Welcome Screen to completion (Enso Bot - Nov 24, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for yesterday (2021-11-24):

**Today:** fixing issues spotted on review, implemented lazy initialization of project view. Preparing PR for interactive functionality of welcome screen, listing projects and opening projects is now implemented. Faced two independent bugs that need some further investigation, will post a description tomorrow. It should be finished by 2021-11-26.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180312485). preparing changes for creating new projects and investigating spotted bugs. (Enso Bot - Nov 25, 2021)
<hr />
**Ilya Bogdanov** reports a new **🔴 BLOCKER** for yesterday (2021-11-25):

**Summary:** The [Developers should be able to work on the welcome screen in our Rust codebase  (#180312485) task](https://www.pivotaltracker.com/story/show/180312485) is blocked by the [#180438664 task](https://www.pivotaltracker.com/story/show/180438664).

see bug description. It significantly affects user experience, so blocks entire Welcome Screen task.

**Possible solutions:** I need to investigate it further, right now there is one possible solution - to disable lazy initialisation of Project View and replace it with immediate initialisation on IDE start. It's easy to implement and won't cause visible performance problems, but might become a blocker in the future. (Enso Bot - Nov 25, 2021)
<hr />
**Ilya Bogdanov** reports a new **🔴 DELAY** for yesterday (2021-11-25):

**Summary:** There is **5 days** delay in implementation of the [Developers should be able to work on the welcome screen in our Rust codebase  (#180312485) task](https://www.pivotaltracker.com/story/show/180312485).
It **will cause 2 days delay** for the delivery of this sprint.

I'm not certain about the delay size, as it caused by several reasons. 2 days of the delay are because of weekends.

**Delay Cause:** my unexpected day off, issue with DOM layers (#180388817), issue with code editor (#180438664) my yet-another-time failed attempts to estimate time needed for review of PRs. As two blockers of the main task (#180388817 and #180388845) are not merged yet, it doesn't look realistic to expect a fully implemented welcome screen to be ready for the release.

**Possible solutions:** We might deliver it behind a feature-flag though, so it won't block further work on IDE startup and initialisation. (Enso Bot - Nov 25, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for yesterday (2021-11-25):

**Today:** mostly investigating stupid welcome-screen layer bug, fixing review comments, finishing up interactive functionality pieces. Opening, listing and creating projects is actually working now. Also faced bug with code editor, separate blocker will be announced. It should be finished by 2021-11-29.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180312485). I will not be available on Friday, but I will spend some time to work with my PRs on the weekends. I need to investigate code editor bug, fix dom layer bug and push my PRs to merge. (Enso Bot - Nov 25, 2021)
<hr />
**Ilya Bogdanov** reports a new **🔴 DELAY** for today (2021-11-29):

**Summary:** There is **4 days** delay in implementation of the [Developers should be able to work on the welcome screen in our Rust codebase  (#180312485) task](https://www.pivotaltracker.com/story/show/180312485).
It **will cause 0 days delay** for the delivery of this sprint.

2 days of the delay are fake and were reported last week, that's because I messed up standup for Thursday back then. Other 2 days are a conservative estimate of the time needed to pass the review.

**Delay Cause:** It takes more time than expected to fix all the small issues with Welcome Screen and to pass a review. Also testing welcome screen changes takes significant time because I have a poor hardware available right now.

**Possible solutions:** latest additions to handbook about the review process should be helpful. I will also return to my normal workplace next week, though I would need to finish Welcome Screen before that time. (Enso Bot - Nov 29, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for today (2021-11-29):

**Today:** I was recording a screencast for PR #3164 (Adding Welcome Screen) and also testing Welcome screen in desktop IDE. I fixed a really annoying bug introduced while working on #180388817. I also tested one of the possible fixes to "pointer-events problem". I found a quick-and-dirty fix for bug #180438664. Given all that, I have a fully working Welcome Screen in a separate branch and just need to test it and get all the changes merged. It should be finished by 2021-12-03.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180312485). I will reopen PR #3159 with necessary fixes, I will try to get PR #3164 merged, I will also prepare PR #3167 with all the missing pieces of the Welcome Screen.

**Additional Information:** I have some performance issues with testing desktop IDE locally, so I might need to find a volunteer to look at it. I will try to use available Windows machine though. (Enso Bot - Nov 29, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for today (2021-11-30):

**Today:** merged two PRs, tested "pointer-events" problem fix, prepared PR for task #180388817, faced some bug with displaying nodes and spend some time reproducing it It should be finished by 2021-12-03.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180312485). preparing PR #3167 with final parts of Welcome Screen (Enso Bot - Nov 30, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for today (2021-12-01):

**Today:** finally tested and confirmed that new `welcome_screen` DOM layer would work correctly. Bisected git history to find a commit responsible for invalid node positions. Rebased and finished a final Welcome Screen PR #3167. It should be finished by 2021-12-03.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180312485). I will make a screen recording of the new Welcome Screen and open #3167 for reviewers. I will merge PR #3171 if no objections will arise.

**Additional Information:** Today I was walking with a dog in the forest (almost got lost a few times). It turns out this dog desperately fears snowmen. (Enso Bot - Dec 1, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for today (2021-12-02):

**Today:** trying to record a screencast with desktop welcome screen. No success because of a poor hardware available at the moment. Wrote documentation for mouse events handling. It should be finished by 2021-12-03.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180312485). hopefully merge the last PR and therefore finish the whole Welcome Screen task (Enso Bot - Dec 2, 2021)
<hr />
**Ilya Bogdanov** reports a new **🔴 DELAY** for today (2021-12-03):

**Summary:** There is **3 days** delay in implementation of the [Developers should be able to work on the welcome screen in our Rust codebase  (#180312485) task](https://www.pivotaltracker.com/story/show/180312485).
It **will cause 0 days delay** for the delivery of this sprint.

2 days of the delay are actually a weekends.

**Delay Cause:** slow PR preparation (due to working with other bugs and issues, like #180388817) lead to a slight delay with a review because all reviewers were quite busy with their tasks.

**Possible solutions:** we'll make a review call on Monday with Michael and Adam. (Enso Bot - Dec 3, 2021)
<hr />
**Ilya Bogdanov** reports a new **🔴 DELAY** for today (2021-12-06):

**Summary:** There is **1 day** delay in implementation of the [Developers should be able to work on the welcome screen in our Rust codebase  (#180312485) task](https://www.pivotaltracker.com/story/show/180312485).
It **will cause 0 days delay** for the delivery of this sprint.

Probably even less than 1 day

**Delay Cause:** another slight delay with Welcome Screen. We made a review with Adam, but I spent some time on setting up dev environment on the new laptop (see #180544972)

**Possible solutions:** report bigger delays early? (Enso Bot - Dec 6, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for today (2021-12-06):

**Today:** Fixed Code Editor bug, made an review call with Adam, fixed spotted issues. I was setting up new M1 MacBook - apparently building IDE does not work at the moment (see #180544972). It should be finished by 2021-12-07.

**Next Day:** Next day I will be working on the [##180303822 task](https://www.pivotaltracker.com/story/show/180303822). merge Welcome Screen and start working on the next issue - fixing warnings in logs. I will probably start with researching available issues. (Enso Bot - Dec 6, 2021)
<hr />
**Ilya Bogdanov** reports a new **🔴 DELAY** for yesterday (2021-12-07):

**Summary:** There is **2 days** delay in implementation of the [Developers should be able to work on the welcome screen in our Rust codebase  (#180312485) task](https://www.pivotaltracker.com/story/show/180312485).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** PR needs one more code owner review from either Michael or Wojciech.

**Possible solutions:** change code owners so that more people would be able to do review? (Enso Bot - Dec 8, 2021)
<hr />
**Ilya Bogdanov** reports a new **🔴 DELAY** for today (2021-12-10):

**Summary:** There is **1 day** delay in implementation of the [Developers should be able to work on the welcome screen in our Rust codebase  (#180312485) task](https://www.pivotaltracker.com/story/show/180312485).
It **will cause 0 day delay** for the delivery of this sprint.

any further delays will affect sprint delivery

**Delay Cause:** slow review made me transfer my attention to the next task, and as the review was done, big architectural changes were requested. (Enso Bot - Dec 9, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for today (2021-12-10):

**Today:** I was doing a massive refactoring of Welcome Screen as proposed by Wojciech. I was a bit confused by compiler errors, didn't understand the simplest way of implementation right away and was studying similar code in the codebase, so it took a lot of time. It should be finished by 2021-12-10.

**Next Day:** Next day I will be working on the [##180560409 task](https://www.pivotaltracker.com/story/show/180560409). I will return to source map warning. (Enso Bot - Dec 9, 2021)
<hr />
**Ilya Bogdanov** reports a new **STANDUP** for the provided date (2021-12-09):

**Today:** I was doing a massive refactoring of Welcome Screen as proposed by Wojciech. I was a bit confused by compiler errors, didn't understand the simplest way of implementation right away and was studying similar code in the codebase, so it took a lot of time. It should be finished by 2021-12-10.

**Next Day:** Next day I will be working on the [##180560409 task](https://www.pivotaltracker.com/story/show/180560409). I will return to source map warning. (Enso Bot - Dec 13, 2021)
<hr />
	@ilyabogdanov there is one regression: when user open the example the name of newly created project is "Unnamed" instead of ie. "Stargazers"  (Sylwia Brodacka - Dec 17, 2021)
<hr />
Regression identified: project created by opening example from welcome screen has name "Unnamed" instead of ie. "Stargazers" (Sylwia Brodacka - Dec 17, 2021)
<hr />
	@sylwia_brodacka 
Fix is on the way in https://github.com/enso-org/enso/pull/3201 (Ilya Bogdanov - Dec 20, 2021)
<hr />
